### ISSUE TITLE
Beyond Zork makes Bocfel exit with object out of range error

### ISSUE THREAD

**Issue by angstsmurf** [garglk/garglk#395]
See https://github.com/angstsmurf/spatterlight/issues/13

```
> blow bubble
You blow gently into the 

fatal error: object 41920 out of range (pc = 0xfd85)
```

The same command works just fine in Frotz and Yazmin:

```
>blow bubble
You blow gently into the circlet's swirling film.
  A silvery bubble blows out of the circlet! You watch as it flattens into a perfectly circular mirror, rotating more and more slowly until it faces southeast.
```

Game version r60, save file attached in the issue linked above.

**Comment by cspiegel** [comment]
This occurs in `@get_prop_addr`, and Frotz, when the game is Beyond Zork, just stores a value of 0 if the object is larger than 2000 (which looks arbitrary).

I suspect Yazmin does something similar.

Zoom crashes, and while I haven't debugged it, I assume it's because the large object number causes an out-of-memory access (and would do so on Bocfel as well if there weren't the assertion).

The primary problem is that there is no explicit maximum object file provided by a game.  If it's outside of dynamic memory, then it's obviously invalid.  The Frotz approach of making up a value of 2000 for the max object number "works", but it's just a band-aid.  It's also a band-aid to simply return 0 for objects outside of dynamic memory.

This does feel like a bug in Beyond Zork, assuming that the Z-machine Standards Document gets objects correct.  Ideally runtime patching would be used, but as of yet I can't figure out what BZ is trying to do here.  I'll see if anybody on intfiction.org can help piece it together.

Worse comes to worst I'll add some sort of upper bound check for BZ as in Frotz.

**Comment by angstsmurf** [comment]
It seems that just commenting out the assertion (in Bocfel) makes it possible to finish the game in both Spatterlight and Gargoyle, but that is obviously not an ideal solution.

**Comment by cspiegel** [comment]
Right, I neglected to think about static memory.  So, the objet *is* in static memory, and thus won't cause a segfault, but objects are required to be in dynamic memory, so it really is an invalid object.  Which means I'm not sure why Zoom crashes, but it does.

If I can't get resolution on why BZ is doing this I'll add a special case to ignore "too-large" objects just in BZ.  But I'd prefer to patch the game to avoid that.  Either way I'll make sure a fix is in for the next release.

**Comment by cspiegel** [comment]
OK, thanks to quick responses from borg323 and Torbjörn Andersson on intfiction, I'm confident I've got a patch that correctly fixes the problem.  I can only test release 60 (since it looks like this occurs late in the game, and 60 is the one a save file was provided for), but it was easy enough to find the same offending code in the disassembly of all releases.

As usual, I'll have this in the next release, but I'll attach the changes here.

[bz.txt](https://github.com/garglk/garglk/files/5587889/bz.txt)

**Comment by angstsmurf** [comment]
Thanks a lot! I've applied this now, and it seems to work fine.

**Comment by cspiegel** [comment]
Apologies, but after some reflection I realized that the bug can actually be fixed instead of papering over it.  The calls can be rewritten to use the proper object instead of just papering over the issue.  I've updated the patches to do this.  If you're interested I'm attaching a patch (against the already-patched version since you've already applied it).  It's not necessary at all since the previous patch does fix the bug, but since I've got fixes instead of band-aids for other games, it seemed wrong to just band-aid this one.

[bz2.txt](https://github.com/garglk/garglk/files/5594388/bz2.txt)
